viz/pkg/healthcheck/healthcheck.go

```diff
@@ -54,147 +72,242 @@ func (hc *HealthChecker) RunChecks(observer healthcheck.CheckObserver) bool {
+					pods, err := hc.KubeAPIClient().GetPodsByNamespace(ctx, hc.vizNamespace)
+					if err != nil {
+						return err
+					}
+					return healthcheck.CheckIfDataPlanePodsExist(pods)
+				}),
+			*healthcheck.NewChecker("viz extension pods are running").
+				WithHintAnchor("l5d-viz-pods-running").
+				Warning().
+				WithRetryDeadline(hc.RetryDeadline).
+				SurfaceErrorOnRetry().
+				WithCheck(func(ctx context.Context) error {
+					pods, err := hc.KubeAPIClient().GetPodsByNamespace(ctx, hc.vizNamespace)
+					if err != nil {
+						return err
+					}
 
-				// Check for relevant pods to be present
-				err = healthcheck.CheckForPods(pods, []string{"linkerd-grafana", "linkerd-prometheus", "linkerd-web", "linkerd-tap"})
-				if err != nil {
-					return err
+					// Check for relevant pods to be present
+					err = healthcheck.CheckForPods(pods, []string{"linkerd-grafana", "linkerd-prometheus", "linkerd-web", "linkerd-tap"})
+					if err != nil {
+						return err
+					}
+
+					return healthcheck.CheckPodsRunning(pods, "")
+				}),
+			*healthcheck.NewChecker("can initialize the client").
+				WithHintAnchor("l5d-viz-existence-client").
+				Fatal().
+				WithCheck(func(ctx context.Context) (err error) {
+					hc.vizAPIClient, err = client.NewExternalClient(ctx, hc.vizNamespace, hc.KubeAPIClient())
+					return
+				}),
+			*healthcheck.NewChecker("data plane proxy metrics are present in Prometheus").
+				WithHintAnchor("l5d-data-plane-prom").
+				Warning().
+				WithRetryDeadline(hc.RetryDeadline).
+				WithCheck(func(ctx context.Context) (err error) {
+					pods, err := hc.getDataPlanePodsFromVizAPI(ctx)
+					if err != nil {
+						return err
+					}
+
+					// TODO: Check if prometheus is present
+
+					return validateDataPlanePodReporting(pods)
+				}),
+			*healthcheck.NewChecker("viz extension self-check").
+				WithHintAnchor("l5d-api-control-api").
+				Fatal().
+				// to avoid confusing users with a prometheus readiness error, we only show
+				// "waiting for check to complete" while things converge. If after the timeout
+				// it still hasn't converged, we show the real error (a 503 usually).
+				WithRetryDeadline(hc.RetryDeadline).
+				WithCheckRPC(func(ctx context.Context) (*healthcheckPb.SelfCheckResponse, error) {
+					return hc.vizAPIClient.SelfCheck(ctx, &healthcheckPb.SelfCheckRequest{})
+				}),
+		}, true),
+
+		*healthcheck.NewCategory(LinkerdVizExtensionDataPlaneCheck, []healthcheck.Checker{
+			*healthcheck.NewChecker("data plane namespace exists").
+				WithHintAnchor("l5d-data-plane-exists").
+				Fatal().
+				WithCheck(func(ctx context.Context) error {
+					if hc.DataPlaneNamespace == "" {
+						// when checking proxies in all namespaces, this check is a no-op
+						return nil
+					}
+					return hc.CheckNamespace(ctx, hc.DataPlaneNamespace, true)
+				}),
+			*healthcheck.NewChecker("data-plane pods have the correct tap configuration").
+				WithHintAnchor("l5d-viz-data-plane-tap").
+				Fatal().
```

Comment:
let's make this non-fatal and make it a warning.  It will be fairly normal for there to be pods which already existed before the tap-injector started up.  it will not be possible to tap those pods until they are restarted, but that's an okay state to be in as long as the tap command provides clear error messages (as is specified in https://github.com/linkerd/linkerd2/issues/5544)